File: src/features/convert/audioOptions.ts
import type {
  AudioBitrate,
  AudioChannels,
  AudioConvertOptions,
  AudioProbeDetails,
  AudioSampleRate,
  AudioSettingsCapabilities,
  GifOptions,
  ImageConvertOptions,
  QualityPreset,
  ResolutionCap,
  SubtitleOptions,
  TargetFormat,
  VideoConvertOptions,
} from "@/types";

export type { AudioBitrate, AudioChannels, AudioConvertOptions, AudioSampleRate };

export type AudioTarget = Extract<
  TargetFormat,
  "mp3" | "m4a" | "aac" | "wav" | "flac"
>;

export interface AudioAvailability {
  available: boolean;
  reason: string | null;
  copyAvailable: boolean;
  copyReason: string | null;
  encodeAvailable: boolean;
  encodeReason: string | null;
  targetCodec?: string;
  bitrateChoicesKbps?: readonly number[];
  defaultBitrateKbps?: number | null;
  defaultChannels?: AudioChannels | null;
  defaultSampleRate?: AudioSampleRate | null;
}

export interface AudioSourceFacts {
  audioStreamCount: number;
  sampleRateHz: number | null;
  channelCount: number | null;
  channelLayoutReported: boolean;
  hasNonAudioStreams: boolean;
}

export interface AudioDraftFile {
  target: TargetFormat;
  audioOptions?: AudioConvertOptions | null;
  audioAvailability?: AudioAvailability | null;
  audioSource?: AudioSourceFacts | null;
  audioBitrateDraft?: string;
  videoOptions?: VideoConvertOptions | null;
  imageOptions?: ImageConvertOptions | null;
  gifOptions?: GifOptions | null;
  compressMode?: unknown;
  subtitle?: SubtitleOptions | null;
  qualityPreset?: QualityPreset | null;
  resolutionCap?: ResolutionCap | null;
}

const AUDIO_TARGETS = new Set<TargetFormat>(["mp3", "m4a", "aac", "wav", "flac"]);
const MP3_BITRATES = [64, 96, 128, 160, 192, 256, 320] as const;
const AAC_BITRATES = [64, 96, 128, 160, 192, 256] as const;
const SAMPLE_RATES = new Set([44_100, 48_000]);
export const audioDraftSlots = ["AudioOptionsPanel.bitrateDraft"] as const;

const object = (value: unknown): value is Record<string, unknown> =>
  value !== null && typeof value === "object" && !Array.isArray(value);
const only = (value: Record<string, unknown>, keys: string[]) =>
  Object.keys(value).every((key) => keys.includes(key));

export function isAudioTarget(target: TargetFormat): target is AudioTarget {
  return AUDIO_TARGETS.has(target);
}

export function audioBitratesForTarget(target: TargetFormat): readonly number[] {
  if (target === "mp3") return MP3_BITRATES;
  if (target === "m4a" || target === "aac") return AAC_BITRATES;
  return [];
}

export function cloneAudioOptions(
  value: AudioConvertOptions | null | undefined,
): AudioConvertOptions | null {
  if (value == null) return null;
  if (value.kind === "copy") return { kind: "copy" };
  return {
    kind: "encode",
    bitrate: value.bitrate ? { ...value.bitrate } : null,
    channels: { ...value.channels },
    sample_rate: { ...value.sample_rate },
  };
}

export function defaultAudioOptions(
  target: TargetFormat,
  availability?: AudioAvailability | null,
): Extract<AudioConvertOptions, { kind: "encode" }> | null {
  if (!isAudioTarget(target)) return null;
  const choices = availability?.bitrateChoicesKbps ?? audioBitratesForTarget(target);
  return {
    kind: "encode",
    bitrate: choices.length ? { kind: "target", kbps: availability?.defaultBitrateKbps ?? 192 } : null,
    channels: { ...(availability?.defaultChannels ?? { kind: "preserve" }) },
    sample_rate: { ...(availability?.defaultSampleRate ?? { kind: "exact", hz: 48_000 }) },
  };
}

export function audioAvailability(
  settings: AudioSettingsCapabilities | null | undefined,
  targetAvailable: boolean,
  targetReason: string | null | undefined,
): AudioAvailability {
  return {
    available: targetAvailable && settings != null,
    reason: targetAvailable ? (settings ? null : "Explicit audio settings are unavailable for this output") : targetReason ?? "This output is unavailable for the source",
    copyAvailable: settings?.copy.available ?? false,
    copyReason: settings?.copy.reason ?? null,
    encodeAvailable: settings?.encode.available ?? false,
    encodeReason: settings?.encode.reason ?? null,
    targetCodec: settings?.target_codec,
    bitrateChoicesKbps: settings?.bitrate_choices_kbps,
    defaultBitrateKbps: settings?.default_bitrate_kbps ?? null,
    defaultChannels: settings?.default_channels ?? null,
    defaultSampleRate: settings?.default_sample_rate ?? null,
  };
}

export function audioSourceFacts(
  details: AudioProbeDetails | null | undefined,
  fallbackAudioStreamCount = 0,
  fallbackHasNonAudioStreams = false,
): AudioSourceFacts {
  const stream = details?.streams[0];
  return {
    audioStreamCount: details?.streams.length ?? fallbackAudioStreamCount,
    sampleRateHz: stream?.sample_rate_hz?.kind === "exact"
      ? stream.sample_rate_hz.value
      : null,
    channelCount: stream?.channels?.kind === "exact" ? stream.channels.value : null,
    channelLayoutReported: Boolean(stream?.channel_layout),
    hasNonAudioStreams: details?.has_non_audio_streams ?? fallbackHasNonAudioStreams,
  };
}

export function audioOptionsForTarget(
  value: AudioConvertOptions | null | undefined,
  target: TargetFormat,
): AudioConvertOptions | null {
  if (!value || !isAudioTarget(target)) return null;
  if (value.kind === "copy") return { kind: "copy" };
  const choices = audioBitratesForTarget(target);
  const bitrate = choices.length === 0
    ? null
    : value.bitrate && choices.includes(value.bitrate.kbps)
      ? { ...value.bitrate }
      : { kind: "target" as const, kbps: 192 };
  return { ...value, bitrate };
}

function validateChannels(value: unknown): AudioChannels {
  if (!object(value) || !only(value, ["kind"])) {
    throw new Error("Audio channels must be Preserve, Mono or Stereo");
  }
  if (value.kind === "preserve" || value.kind === "mono" || value.kind === "stereo") {
    return { kind: value.kind };
  }
  throw new Error("Audio channels must be Preserve, Mono or Stereo");
}

function validateSampleRate(value: unknown): AudioSampleRate {
  if (!object(value)) {
    throw new Error("Audio sample rate must be Preserve, 44.1 kHz or 48 kHz");
  }
  if (value.kind === "preserve" && only(value, ["kind"])) {
    return { kind: "preserve" };
  }
  if (
    value.kind === "exact" &&
    only(value, ["kind", "hz"]) &&
    typeof value.hz === "number" &&
    Number.isSafeInteger(value.hz) &&
    SAMPLE_RATES.has(value.hz)
  ) {
    return { kind: "exact", hz: value.hz };
  }
  throw new Error("Audio sample rate must be Preserve, 44.1 kHz or 48 kHz");
}

export function validateAudioOptions(value: unknown): AudioConvertOptions | null {
  if (value == null) return null;
  if (object(value) && value.kind === "copy" && only(value, ["kind"])) {
    return { kind: "copy" };
  }
  if (
    !object(value) ||
    value.kind !== "encode" ||
    !only(value, ["kind", "bitrate", "channels", "sample_rate"])
  ) {
    throw new Error("Audio options must specify Automatic, Copy or complete Custom encode settings");
  }
  let bitrate: AudioBitrate | null = null;
  if (value.bitrate != null) {
    if (
      !object(value.bitrate) ||
      value.bitrate.kind !== "target" ||
      !only(value.bitrate, ["kind", "kbps"]) ||
      typeof value.bitrate.kbps !== "number" ||
      !Number.isSafeInteger(value.bitrate.kbps)
    ) {
      throw new Error("Audio bitrate must be one of the target's supported choices");
    }
    bitrate = { kind: "target", kbps: value.bitrate.kbps };
  }
  return {
    kind: "encode",
    bitrate,
    channels: validateChannels(value.channels),
    sample_rate: validateSampleRate(value.sample_rate),
  };
}

type AudioRequest = {
  target: TargetFormat;
  audio_options?: unknown;
  video_options?: unknown;
  image_options?: unknown;
  gif_options?: unknown;
  compress_mode?: unknown;
  subtitle?: unknown;
  quality_preset?: unknown;
  resolution_cap?: unknown;
};

export function validateAudioRequest(request: AudioRequest): AudioConvertOptions | null {
  const options = validateAudioOptions(request.audio_options);
  if (!options) return null;
  if (!isAudioTarget(request.target)) {
    throw new Error("Explicit audio settings require MP3, M4A, AAC, WAV or FLAC output");
  }
  if (
    request.video_options != null ||
    request.image_options != null ||
    request.gif_options != null ||
    request.compress_mode != null ||
    request.subtitle != null ||
    request.quality_preset != null ||
    request.resolution_cap != null
  ) {
    throw new Error("Explicit audio settings cannot be combined with video, image, GIF, compression, subtitle, quality or resolution settings");
  }
  if (options.kind === "encode") {
    const choices = audioBitratesForTarget(request.target);
    if (choices.length === 0 && options.bitrate !== null) {
      throw new Error(`${request.target.toUpperCase()} Custom encode does not use a bitrate target`);
    }
    if (choices.length > 0 && (!options.bitrate || !choices.includes(options.bitrate.kbps))) {
      throw new Error(`${request.target.toUpperCase()} bitrate must be one of ${choices.join(", ")} kbps`);
    }
  }
  return cloneAudioOptions(options);
}

export function audioOptionsProblem(file: AudioDraftFile): string | null {
  if (!file.audioOptions) return null;
  try {
    validateAudioRequest({
      target: file.target,
      audio_options: file.audioOptions,
      video_options: file.videoOptions,
      image_options: file.imageOptions,
      gif_options: file.gifOptions,
      compress_mode: file.compressMode,
      subtitle: file.subtitle,
      quality_preset: file.qualityPreset,
      resolution_cap: file.resolutionCap,
    });
  } catch (error) {
    return error instanceof Error ? error.message : String(error);
  }
  if (file.audioOptions.kind === "encode" && audioBitratesForTarget(file.target).length > 0) {
    const raw = file.audioBitrateDraft ?? String(file.audioOptions.bitrate?.kbps ?? "");
    if (!/^\d+$/.test(raw) || !audioBitratesForTarget(file.target).includes(Number(raw))) {
      return `${file.target.toUpperCase()} bitrate must be one of ${audioBitratesForTarget(file.target).join(", ")} kbps`;
    }
  }
  if (file.audioSource && file.audioSource.audioStreamCount !== 1) {
    return "Explicit audio settings need exactly one audio track. Track selection is not available yet.";
  }
  if (file.audioAvailability) {
    if (!file.audioAvailability.available) {
      return file.audioAvailability.reason ?? "Audio settings are unavailable for this output";
    }
    if (file.audioOptions.kind === "copy" && !file.audioAvailability.copyAvailable) {
      return file.audioAvailability.copyReason ?? "Copy audio is unavailable for this source";
    }
    if (file.audioOptions.kind === "encode" && !file.audioAvailability.encodeAvailable) {
      return file.audioAvailability.encodeReason ?? "Custom audio encoding is unavailable for this source";
    }
  }
  if (
    file.audioOptions.kind === "encode" &&
    file.audioOptions.sample_rate.kind === "preserve" &&
    file.audioSource &&
    !SAMPLE_RATES.has(file.audioSource.sampleRateHz ?? 0)
  ) {
    return "Preserve sample rate requires a reported 44.1 or 48 kHz source. Choose an exact rate.";
  }
  return null;
}

export function audioRequestOptions(file: AudioDraftFile): AudioConvertOptions | null {
  const problem = audioOptionsProblem(file);
  if (problem) throw new Error(problem);
  return validateAudioRequest({
    target: file.target,
    audio_options: cloneAudioOptions(file.audioOptions),
    video_options: file.videoOptions,
    image_options: file.imageOptions,
    gif_options: file.gifOptions,
    compress_mode: file.compressMode,
    subtitle: file.subtitle,
    quality_preset: file.qualityPreset,
    resolution_cap: file.resolutionCap,
  });
}

File: crates/goop-converter/src/audio_options.rs
use crate::{compat::Plan, encoders::DetectedEncoders};
use goop_core::{
    validate_audio_request, AudioBitrate, AudioChannels, AudioConvertOptions,
    AudioExecutionSummary, AudioModeAvailability, AudioNumericFact, AudioSampleRate,
    AudioSettingsCapabilities, AudioStreamInfo, ConvertRequest, GoopError, ProbeResult,
    TargetFormat, VideoRationalFact,
};

#[derive(Debug, Clone)]
pub struct ResolvedAudio {
    pub plan: Plan,
    pub summary: AudioExecutionSummary,
}

fn invalid(message: impl Into<String>) -> GoopError {
    GoopError::InvalidRequest(message.into())
}

fn exact(fact: &Option<AudioNumericFact>) -> Option<u32> {
    match fact {
        Some(AudioNumericFact::Exact { value }) => Some(*value),
        _ => None,
    }
}

fn target_facts(target: TargetFormat) -> Option<(&'static str, &'static str, &'static [u32])> {
    match target {
        TargetFormat::Mp3 => Some(("mp3", "libmp3lame", &[64, 96, 128, 160, 192, 256, 320])),
        TargetFormat::M4a | TargetFormat::Aac => {
            Some(("aac", "aac", &[64, 96, 128, 160, 192, 256]))
        }
        TargetFormat::Wav => Some(("pcm_s16le", "pcm_s16le", &[])),
        TargetFormat::Flac => Some(("flac", "flac", &[])),
        _ => None,
    }
}

fn source_stream(probe: &ProbeResult) -> Result<&AudioStreamInfo, GoopError> {
    let details = probe.audio_details.as_ref().ok_or_else(|| {
        invalid("Explicit audio settings require complete fresh audio stream facts")
    })?;
    if details.streams.len() != 1 {
        return Err(invalid(
            "Choose Automatic until audio-track selection is available",
        ));
    }
    Ok(&details.streams[0])
}

fn source_stream_for_request<'a>(
    request: &ConvertRequest,
    probe: &'a ProbeResult,
) -> Result<&'a AudioStreamInfo, GoopError> {
    if request.track_options.is_some() {
        crate::track_options::selected_audio_stream(request, probe)
    } else {
        source_stream(probe)
    }
}

#[derive(Debug)]
struct ValidatedAudio<'a> {
    copied: bool,
    sample_rate_hz: u32,
    channels: u32,
    channel_layout: Option<&'a str>,
    sample_format: Option<&'a str>,
    bit_depth: Option<u32>,
}

fn validate_options_for_stream<'a>(
    requested: &AudioConvertOptions,
    source: &'a AudioStreamInfo,
    target: TargetFormat,
    encoders: &DetectedEncoders,
) -> Result<ValidatedAudio<'a>, GoopError> {
    let (codec, encoder, _) = target_facts(target)
        .ok_or_else(|| invalid("This output does not support explicit audio settings"))?;
    let source_channels = exact(&source.channels).filter(|value| *value > 0);
    let source_rate = exact(&source.sample_rate_hz);
    match requested {
        AudioConvertOptions::Copy => {
            if source.codec_name.as_deref() != Some(codec) {
                return Err(invalid(format!(
                    "Copy audio requires {codec} source audio for this output"
                )));
            }
            let sample_rate_hz = source_rate
                .filter(|value| *value > 0)
                .ok_or_else(|| invalid("Copy audio requires a reported sample rate"))?;
            let channels = source_channels
                .ok_or_else(|| invalid("Copy audio requires a reported channel count"))?;
            Ok(ValidatedAudio {
                copied: true,
                sample_rate_hz,
                channels,
                channel_layout: source.channel_layout.as_deref(),
                sample_format: source.sample_format.as_deref(),
                bit_depth: exact(&source.bits_per_raw_sample).filter(|value| *value > 0),
            })
        }
        AudioConvertOptions::Encode {
            channels,
            sample_rate,
            ..
        } => {
            let source_channels = source_channels
                .filter(|value| matches!(value, 1 | 2))
                .ok_or_else(|| {
                    invalid(
                        "Custom audio settings require a source with exactly one or two reported channels",
                    )
                })?;
            if !encoders.is_available(encoder) {
                return Err(invalid(format!(
                    "Custom audio encoding requires the {encoder} encoder"
                )));
            }
            let channel_layout = match channels {
                AudioChannels::Preserve => source.channel_layout.as_deref(),
                AudioChannels::Mono => Some("mono"),
                AudioChannels::Stereo => Some("stereo"),
            };
            let channels = match channels {
                AudioChannels::Preserve => source_channels,
                AudioChannels::Mono => 1,
                AudioChannels::Stereo => 2,
            };
            let sample_rate_hz = match sample_rate {
                AudioSampleRate::Preserve => source_rate
                    .filter(|value| matches!(value, 44_100 | 48_000))
                    .ok_or_else(|| {
                        invalid(
                            "Preserve sample rate requires reported 44100 or 48000 Hz source audio",
                        )
                    })?,
                AudioSampleRate::Exact { hz } => *hz,
            };
            Ok(ValidatedAudio {
                copied: false,
                sample_rate_hz,
                channels,
                channel_layout,
                sample_format: None,
                bit_depth: (target == TargetFormat::Wav).then_some(16),
            })
        }
    }
}

pub fn resolve(
    request: &ConvertRequest,
    probe: &ProbeResult,
    encoders: &DetectedEncoders,
) -> Result<ResolvedAudio, GoopError> {
    validate_audio_request(request)?;
    let requested = request
        .audio_options
        .as_ref()
        .ok_or_else(|| invalid("Explicit audio settings are missing"))?;
    let (codec, encoder, _) = target_facts(request.target)
        .ok_or_else(|| invalid("This output does not support explicit audio settings"))?;
    let source = source_stream_for_request(request, probe)?;
    let validated = validate_options_for_stream(requested, source, request.target, encoders)?;
    let notice = probe
        .audio_details
        .as_ref()
        .is_some_and(|details| details.has_non_audio_streams && request.track_options.is_none())
        .then(|| "Non-audio streams and artwork are not included".to_owned());

    let mut args = vec![
        "-map".into(),
        format!("0:{}", source.index),
        "-vn".into(),
        "-sn".into(),
        "-dn".into(),
    ];
    match requested {
        AudioConvertOptions::Copy => args.extend(["-c:a".into(), "copy".into()]),
        AudioConvertOptions::Encode {
            bitrate,
            channels,
            sample_rate,
        } => {
            args.extend(["-c:a".into(), encoder.into()]);
            if let Some(AudioBitrate::Target { kbps }) = bitrate {
                args.extend(["-b:a".into(), format!("{kbps}k")]);
            }
            match channels {
                AudioChannels::Preserve => {}
                AudioChannels::Mono if exact(&source.channels) == Some(2) => {
                    args.extend(["-af".into(), "pan=mono|c0=0.5*c0+0.5*c1".into()]);
                }
                AudioChannels::Mono => args.extend(["-ac".into(), "1".into()]),
                AudioChannels::Stereo => args.extend(["-ac".into(), "2".into()]),
            }
            if let AudioSampleRate::Exact { hz } = sample_rate {
                args.extend(["-ar".into(), hz.to_string()]);
            }
        }
    }

    Ok(ResolvedAudio {
        plan: Plan {
            args,
            video_filters: vec![],
            reencoded: !validated.copied,
            ext: request.target.extension(),
        },
        summary: AudioExecutionSummary {
            requested: requested.clone(),
            encoder: (!validated.copied).then(|| encoder.to_owned()),
            codec: codec.into(),
            audio_stream_index: source.index,
            copied: validated.copied,
            sample_rate_hz: validated.sample_rate_hz,
            channels: validated.channels,
            channel_layout: validated.channel_layout.map(str::to_owned),
            sample_format: validated.sample_format.map(str::to_owned),
            bit_depth: validated.bit_depth,
            reported_bitrate_kbps: if validated.copied {
                exact(&source.bit_rate_bps).map(|bps| bps / 1_000)
            } else {
                None
            },
            notices: notice.into_iter().collect(),
        },
    })
}

fn availability<T>(result: Result<T, GoopError>) -> AudioModeAvailability {
    match result {
        Ok(_) => AudioModeAvailability {
            available: true,
            reason: None,
        },
        Err(error) => AudioModeAvailability {
            available: false,
            reason: Some(error.user_message()),
        },
    }
}

pub fn capabilities(
    probe: &ProbeResult,
    target: TargetFormat,
    encoders: &DetectedEncoders,
) -> AudioSettingsCapabilities {
    capabilities_for_source(source_stream(probe), target, encoders)
}

pub(crate) fn mode_availability_for_stream(
    source: Option<&AudioStreamInfo>,
    target: TargetFormat,
    encoders: &DetectedEncoders,
) -> (AudioModeAvailability, AudioModeAvailability) {
    let (_, _, bitrates) = target_facts(target).unwrap_or(("unsupported", "unsupported", &[]));
    let evaluated = evaluate_capabilities(
        source.ok_or_else(|| {
            invalid("Explicit audio settings require complete fresh audio stream facts")
        }),
        target,
        encoders,
        bitrates,
    );
    (evaluated.copy, evaluated.encode)
}

struct EvaluatedCapabilities<'a> {
    source: Option<&'a AudioStreamInfo>,
    copy: AudioModeAvailability,
    encode: AudioModeAvailability,
    default_channels: Option<AudioChannels>,
    default_sample_rate: Option<AudioSampleRate>,
}

fn evaluate_capabilities<'a>(
    source: Result<&'a AudioStreamInfo, GoopError>,
    target: TargetFormat,
    encoders: &DetectedEncoders,
    bitrates: &[u32],
) -> EvaluatedCapabilities<'a> {
    let (source, source_error) = match source {
        Ok(source) => (Some(source), None),
        Err(error) => (None, Some(error.user_message())),
    };
    let source_rate = source.and_then(|stream| exact(&stream.sample_rate_hz));
    let source_channels = source.and_then(|stream| exact(&stream.channels));
    let default_sample_rate = if matches!(source_rate, Some(44_100 | 48_000)) {
        Some(AudioSampleRate::Preserve)
    } else {
        Some(AudioSampleRate::Exact { hz: 48_000 })
    };
    let default_channels =
        matches!(source_channels, Some(1 | 2)).then_some(AudioChannels::Preserve);
    let encode = AudioConvertOptions::Encode {
        bitrate: (!bitrates.is_empty()).then_some(AudioBitrate::Target { kbps: 192 }),
        channels: default_channels.unwrap_or(AudioChannels::Preserve),
        sample_rate: default_sample_rate.unwrap_or(AudioSampleRate::Exact { hz: 48_000 }),
    };
    let unavailable = |reason: &str| AudioModeAvailability {
        available: false,
        reason: Some(reason.to_owned()),
    };
    let copy = match source {
        Some(source) => availability(validate_options_for_stream(
            &AudioConvertOptions::Copy,
            source,
            target,
            encoders,
        )),
        None => unavailable(
            source_error
                .as_deref()
                .unwrap_or("Audio facts are unavailable"),
        ),
    };
    let encode_availability = match source {
        Some(source) => availability(validate_options_for_stream(
            &encode, source, target, encoders,
        )),
        None => unavailable(
            source_error
                .as_deref()
                .unwrap_or("Audio facts are unavailable"),
        ),
    };
    EvaluatedCapabilities {
        source,
        copy,
        encode: encode_availability,
        default_channels,
        default_sample_rate,
    }
}

fn capabilities_for_source(
    source: Result<&AudioStreamInfo, GoopError>,
    target: TargetFormat,
    encoders: &DetectedEncoders,
) -> AudioSettingsCapabilities {
    let (codec, encoder, bitrates) =
        target_facts(target).unwrap_or(("unsupported", "unsupported", &[]));
    let evaluated = evaluate_capabilities(source, target, encoders, bitrates);
    AudioSettingsCapabilities {
        copy: evaluated.copy,
        encode: evaluated.encode,
        target_codec: codec.into(),
        encoder: encoder.into(),
        bitrate_choices_kbps: bitrates.to_vec(),
        default_bitrate_kbps: (!bitrates.is_empty()).then_some(192),
        channel_choices: vec![
            AudioChannels::Preserve,
            AudioChannels::Mono,
            AudioChannels::Stereo,
        ],
        default_channels: evaluated.default_channels,
        sample_rate_choices: vec![
            AudioSampleRate::Preserve,
            AudioSampleRate::Exact { hz: 44_100 },
            AudioSampleRate::Exact { hz: 48_000 },
        ],
        default_sample_rate: evaluated.default_sample_rate,
        source: evaluated.source.cloned(),
    }
}

pub fn validate_output_against_source(
    expected: &AudioExecutionSummary,
    source: &ProbeResult,
    actual: &ProbeResult,
) -> Result<AudioExecutionSummary, GoopError> {
    let source_audio = source
        .audio_details
        .as_ref()
        .and_then(|details| {
            details
                .streams
                .iter()
                .find(|stream| stream.index == expected.audio_stream_index)
        })
        .ok_or_else(|| invalid("Admitted source audio facts are no longer complete"))?;
    let stream = source_stream(actual).map_err(|_| {
        invalid("Completed audio output did not contain exactly one verified audio stream")
    })?;
    if actual
        .audio_details
        .as_ref()
        .is_some_and(|details| details.has_non_audio_streams)
    {
        return Err(invalid(
            "Completed audio output contains an unexpected non-audio stream",
        ));
    }
    let rate = exact(&stream.sample_rate_hz)
        .ok_or_else(|| invalid("Completed audio output has no verified sample rate"))?;
    let channels = exact(&stream.channels)
        .ok_or_else(|| invalid("Completed audio output has no verified channel count"))?;
    if stream.codec_name.as_deref() != Some(expected.codec.as_str())
        || rate != expected.sample_rate_hz
        || channels != expected.channels
    {
        return Err(invalid(
            "Completed audio output does not match the admitted codec, sample rate and channels",
        ));
    }

    let source_rate = exact(&source_audio.sample_rate_hz).filter(|value| *value > 0);
    let codec_allowance = |codec: Option<&str>, sample_rate: Option<u32>| -> u64 {
        let frames: u64 = match codec {
            Some("aac") => 2 * 1_024,
            Some("mp3") => 2 * 1_152,
            _ => 1,
        };
        sample_rate
            .map(|value| frames * 1_000_u64 / u64::from(value) + 1)
            .unwrap_or(0)
    };
    let tick_allowance = |fact: &Option<VideoRationalFact>, label: &str| match fact {
        Some(VideoRationalFact::Exact {
            numerator,
            denominator,
        }) if *numerator > 0 && *denominator > 0 => {
            Ok(u64::from(*numerator) * 1_000_u64 / u64::from(*denominator) + 1)
        }
        _ => Err(invalid(format!("{label} has no verified stream time base"))),
    };
    let source_duration = source_audio
        .duration_ms
        .filter(|duration| *duration > 0)
        .ok_or_else(|| invalid("Admitted source audio has no verified stream duration"))?;
    let output_duration = stream
        .duration_ms
        .filter(|duration| *duration > 0)
        .ok_or_else(|| invalid("Completed audio output has no verified stream duration"))?;
    let allowance = codec_allowance(source_audio.codec_name.as_deref(), source_rate)
        + codec_allowance(Some(expected.codec.as_str()), Some(rate))
        + tick_allowance(&source_audio.time_base, "Admitted source audio")?
        + tick_allowance(&stream.time_base, "Completed audio output")?;
    if source_duration.abs_diff(output_duration) > allowance {
        return Err(invalid(
            "Completed audio output duration is outside the codec allowance",
        ));
    }

    let mut completed = expected.clone();
    completed.sample_rate_hz = rate;
    completed.channels = channels;
    completed.channel_layout = stream.channel_layout.clone();
    completed.sample_format = stream.sample_format.clone();
    completed.bit_depth = exact(&stream.bits_per_raw_sample).filter(|value| *value > 0);
    completed.reported_bitrate_kbps = exact(&stream.bit_rate_bps).map(|bps| bps / 1_000);
    Ok(completed)
}
